src/Bounds.cpp

```diff
@@ -1530,33 +1552,128 @@ class Bounds : public IRVisitor {
                 result.include(arg_bounds.get(i));
             }
             interval = result;
-        } else if (op->is_intrinsic(Call::halving_add)) {
-            // lower_halving_add() uses bitwise tricks that are hard to reason
-            // about; let's do this instead:
-            Expr e;
-            if (op->type.bits() == 64) {
-                bounds_of_type(t);
-            } else {
-                Expr e = narrow((widen(op->args[0]) + widen(op->args[1])) / 2);
-                e.accept(this);
-            }
-        } else if (op->is_intrinsic(Call::rounding_halving_add)) {
-            // lower_rounding_halving_add() uses bitwise tricks that are hard to reason
-            // about; let's do this instead:
-            if (op->type.bits() == 64) {
-                bounds_of_type(t);
+        } else if (op->is_intrinsic(Call::widen_right_add)) {
```

Comment:
I don't think we need this many safety checks for the widening operations. Any `Expr` in a widening op needs to be able to be widened - we can't lift to `widening_mul` unless a user widened the inputs. We only need to be careful with operations that we can lift to without widening operations, but that the "simple" lowering pattern involves widening.

Reply:
Well, yes and no -- it's true a that the Exprs in a widening op need to be widened, and well-formed code shouldn't pass us cases that don't fit; that said, we absolutely *will* get misuse in that way, so what should we do when that happens? IMHO we are better off checking for it an explicitly devolving to bounds-of-type, rather than risking that the bounds-calc code makes a mistake and calculates an inappropriate bound due to inadvertent overflow.

Reply:
Ah, I think I misunderstood the use case. Is this for when users write code that produces a LUT index, and uses intermediate 64 bit types?

Reply:
> Is this for when users write code that produces a LUT index

Yes? I mean, we have no control of what the user is doing with these functions; they could pass it insane nonsense, so we need to be somewhat defensive here. We'd prefer to avoid a too-loose bounds, but we absolutely cannot risk getting too-tight bounds.